packages/nouns-api/src/services/ideas.ts

```diff
@@ -73,19 +74,30 @@ class IdeasService {
     date,
   }: {
     sortBy?: string;
-    tags?: string[];
+    tags?: TagType[];
     date?: string;
   }) {
     try {
       const dateRange: any = DATE_FILTERS[date || 'ALL_TIME'].filterFn();
       const ideas = await prisma.idea.findMany({
         where: {
+          ...(tags &&
```

Comment:
I think this errors here when we apply a Virtual Tag as it isn't defined as an enum in the prisma schema. I think we can resolve that by filtering those tags out here, maybe something like this?

```
const tagMap = { dbTags: [], virtualTags: [] };
const virtualTags = Object.keys(VirtualTags);
tags.forEach(tag => {
  if (virtualTags.includes(tag)) {
    tagMap.virtualTags.push(tag)
  } else {
    tagMap.dbTags.push(tag)
  }
});
```

You could then use the dbTags in the query and the virtual tags filterFn below where we're building the votecount (around line 114)

Reply:
Yes good catch, although I think all of the tags have to be filtered outside of the db level. The virtual tags are already filtered outside of the db, so we need to filter the regular tags as well outside of the db or else we might prematurely filter some ideas that meet a virtual tag but not a regular tag.

For example, imagine an idea with the tag "HOT" but nothing else.
If we filtered first for regular tags at the db level, this would get filtered out, because there are no matches.

So instead we have to return all of the ideas, then filter outside of the db on both virtual OR regular tags.

Reply:
> For example, imagine an idea with the tag "HOT" but nothing else.
If we filtered first for regular tags at the db level, this would get filtered out, because there are no matches.

Hmm but in this example nothing would get filtered out at the db level as no regular tags would be applied so by default we'd be returning all ideas. When no tags are applied the query should return everything, we can then run the virtual tag filtering on that data set.

Checkout this gist as an example, i've added some comments at each of the steps.

https://gist.github.com/ThatTobMate/8fd01b5302e10d2effb635319fece3ce

With the gist above I think if only `HOT` was applied then `tagMap.dbTags` would be empty so no tags would be applied to the query and we'd get ALL ideas back.
We would then iterate over all ideas, add all the Virtual tags, then filter the ideas out that don't have the `HOT` tag when generating the `ideaData`.

I might be missing something though so let me know if that sounds wrong.

Reply:
Saying that i can see this issue occur where you may apply a virtual tag and a regular tag e.g.  `HOT` and `DISCUSSION`, should the tags be applied as AND or OR. If it's OR then your solution is the only way, if AND then we shuld still be able to have db level filtering too.

Either way happy to keep it as you have it now and we can always change it later if needed

Reply:
> Hmm but in this example nothing would get filtered out at the db level as no regular tags would be applied so by default we'd be returning all ideas. When no tags are applied the query should return everything, we can then run the virtual tag filtering on that data set.

In this case, there might be regular tag filters applied, but our tag in question has only virtual tags.

Here's a bit of an example
Idea 1: REQUEST
Idea 2: HOT
Idea 3: GOVERNANCE

Lets say we apply filters REQUEST and HOT
If we first filter by the non-virtual tags at the db level, we are going to only return idea 1, since HOT is not a db level tag and so the prisma query to request all ideas where tag is in the list of db level tags will only return idea 1. 

Now by the time we are filtering with the virtual tag filters, it will be too late, since Idea 2 was already filtered out incorrectly.

I'll check your gist though, maybe I'm misunderstanding

Reply:
Yep sorry i clarified after that if you mix a regular and virtual tag then we would get into this situation.

> I think even if the logic is AND instead of OR this issue would still present itself.

If it's an AND then we wouldn't as we could filter out at the regular tag level first as we'd know we only ever want ideas with at least those tags.

e.g. In your example above:

- If AND then we could filter at db level to find ideas with `Request` = idea 1 returned, then filter through for ideas with HOT on that and we'd get 0 ideas because no ideas match both filters. If there was an Idea 4 with Request & Hot tags then we would get Idea 4 returned. 
- If OR then we can't filter at db level as ideas 1 and 2 meet the criteria but db level filtering would exclude idea 2 as you mentioned

I think you have it setup as an OR right now which is fine and again we can always change this later if we want to go to AND logic.